File: problem_46.java
import java.util.*;

public class problem_46 {
    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries) {
        // Step 1: Build the graph
        Map<String, Map<String, Double>> graph = new HashMap<>();
        for (int i = 0; i < equations.size(); i++) {
            String a = equations.get(i).get(0);
            String b = equations.get(i).get(1);
            double val = values[i];

            graph.putIfAbsent(a, new HashMap<>());
            graph.putIfAbsent(b, new HashMap<>());
            graph.get(a).put(b, val);
            graph.get(b).put(a, 1.0 / val);
        }

        // Step 2: Process each query
        double[] result = new double[queries.size()];
        for (int i = 0; i < queries.size(); i++) {
            String start = queries.get(i).get(0);
            String end = queries.get(i).get(1);
            Set<String> visited = new HashSet<>();
            result[i] = dfs(start, end, graph, visited);
        }
        return result;
    }

    private double dfs(String start, String end, Map<String, Map<String, Double>> graph, Set<String> visited) {
        if (!graph.containsKey(start) || !graph.containsKey(end)) return -1.0;
        if (graph.get(start).containsKey(end)) return graph.get(start).get(end);

        visited.add(start);

        for (Map.Entry<String, Double> neighbor : graph.get(start).entrySet()) {
            if (!visited.contains(neighbor.getKey())) {
                double temp = dfs(neighbor.getKey(), end, graph, visited);
                if (temp != -1.0) {
                    return neighbor.getValue() * temp;
                }
            }
        }
        return -1.0;
    }

    // Example usage
    public static void main(String[] args) {
        problem_46 solver = new problem_46();

        List<List<String>> equations = Arrays.asList(
                Arrays.asList("a", "b"),
                Arrays.asList("b", "c")
        );
        double[] values = {2.0, 3.0};
        List<List<String>> queries = Arrays.asList(
                Arrays.asList("a", "c"),
                Arrays.asList("b", "a"),
                Arrays.asList("a", "e"),
                Arrays.asList("a", "a"),
                Arrays.asList("x", "x")
        );

        double[] results = solver.calcEquation(equations, values, queries);
        System.out.println(Arrays.toString(results));
    }
}

File: problem_46.py
"""
Problem: Evaluate Division

You are given an array of equations `equations` where each equation is a list of two variables [A, B],
and an array `values` where `values[i]` represents the result of the equation A / B = values[i].
You are also given some `queries`, where each query is a list of two variables [C, D].
Return the answers to each query. If the answer does not exist, return -1.0.

Example:
Input:
equations = [["a","b"],["b","c"]]
values = [2.0,3.0]
queries = [["a","c"],["b","a"],["a","e"],["a","a"],["x","x"]]

Output: [6.0,0.5,-1.0,1.0,-1.0]
"""

import collections
from typing import List

class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        # STEP 1: BUILD THE GRAPH
        graph = collections.defaultdict(dict)
        for (x, y), val in zip(equations, values):
            graph[x][y] = val
            graph[y][x] = 1.0 / val

        # STEP 2: DFS FUNCTION
        def dfs(x, y, visited):
            if x not in graph or y not in graph:
                return -1.0
            if y in graph[x]:
                return graph[x][y]
            for nei in graph[x]:
                if nei not in visited:
                    visited.add(nei)
                    temp = dfs(nei, y, visited)
                    if temp != -1:
                        return graph[x][nei] * temp
            return -1.0

        # STEP 3: PROCESS EACH QUERY
        res = []
        for query in queries:
            res.append(dfs(query[0], query[1], set([query[0]])))
        return res


# Example usage
if __name__ == "__main__":
    equations = [["a", "b"], ["b", "c"]]
    values = [2.0, 3.0]
    queries = [["a", "c"], ["b", "a"], ["a", "e"], ["a", "a"], ["x", "x"]]
    sol = Solution()
    print(sol.calcEquation(equations, values, queries))
